### ISSUE TITLE
Should instance From a a exist?

### ISSUE THREAD

**Issue by 21it** [tfausak/witch#46]
Technically `instance From a a` do fit witch laws, and it's just `id`. Often after code refactoring in apps where `witch` is used, I do find absolutely redundant `from x` expressions. Before refactoring is was using `From a b` instance where `a /= b`, but after refactoring it became `From a a` because type of `x` changed. GHC and hlint are not helpful in this particular case, and job of careful code review to remove redundant code in this case is completely manual. Maybe `instance From a a` just should not exist to make code refactoring simpler? This instance is not very useful anyway.

**Comment by tfausak** [comment]
Great question! I go back and forth on this myself. On the one hand, the `From a a` instance abides by the laws and makes sense as a sort of "identity" instance. On the other hand, I've seen the same thing where I end up with unnecessary `from` calls after refactoring. 

I would love it if there was a way to add a warning about this. As you mentioned, HLint can't quite pull this off since you need information about the type checked program and HLint only deals with syntax. Presumably a GHC plugin could look for expressions like `from @a @a` for any `a`, but I don't know how to write such a plugin currently. 

For posterity, see #27 for where I added `From a a` in the first place.

**Comment by mitchellwrosen** [comment]
Without the `From a a` instance, you cannot write `From`-polymorphic helpers, such as

```haskell
appendBangBang :: From a String => a -> String
appendBangBang x = from @a x ++ "!!"
```

🤷‍♂️

**Comment by 21it** [comment]
I'm not an expert, but it feels like it's still possible without `From a a` instance, but requires a bit of type-level work (idea is widely used for example in singletons library). Some sort of runtime pattern matching of `Refl` which gives type checker a compile-time proof about type `a` and `b` equality or non-equality. For example such technique is used to write `Eq` instances for existentially quantified types like here: https://github.com/21it/src/blob/e2e2834493fcd9e6082e1f95114c36e1b624f328/bitfinex-client/src/BitfinexClient/Data/Metro.hs#L295-L299

**Comment by 21it** [comment]
But I'm not 100% sure yet, I should try this technique with `witch` myself.

**Comment by mitchellwrosen** [comment]
I didn't totally follow that, sorry. But I also wasn't very clear - you _can_ write a function like `appendBangBang`, but the intention of this type of function with a polymorphic input is to allow the caller to pick any `a` that can be converted to `String`. It seems wrong to disallow `String` itself to be picked, for callers that don't need any conversion.

**Comment by tfausak** [comment]
That's a good motivation, @mitchellwrosen! Thanks for providing it. 

Although I wrote Witch, the way that I use it is potentially atypical. I mostly use it at work to convert between various domain specific types and their primitive counterparts. For example `from @Int @UserId 123` type of stuff. 

As a result, I don't write many functions with `From` constraints. But the example you provided makes total sense! I want to continue allowing functions like that.

**Comment by mitchellwrosen** [comment]
Totally! I too think calling `From` at known types is the way to go. But I've occasionally seen helpers that try to convert on behalf of callers, either on the input or output side, which does seem like an okay idea at first, but also tends to go wrong eventually.

**Comment by 21it** [comment]
Here is an example of polymorphic `appendBangBang` which works without `From a a` or `From String String` instances, but still is able to accept `String` value as an argument. I'm not an expert in type-level programming, so implementation looks ugly, but I'm sure it's possible to make it much better, generalize it, hide details, provide simple utilities etc:

```haskell
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE NoImplicitPrelude #-}

module Lib
  ( str,
    txt,
  )
where

import Data.Bool.Singletons
import Data.List.Singletons
import qualified Data.Text as T
import Data.Type.Equality
import Data.Typeable
import Universum

class From a b where
  from :: a -> b

instance From Void b where
  from =
    error $
      "This will never happen"
        <> " because Void values"
        <> " are impossible."

instance From Text String where
  from = T.unpack

instance From String Text where
  from = T.pack

appendBangBang ::
  forall a b.
  ( Typeable a,
    Typeable b,
    If (a == String) (Void ~ b) (a ~ b),
    From b String
  ) =>
  a ->
  String
appendBangBang x =
  case eqT @a @String of
    -- It's a proof for GHC which allows
    -- to apply ++ without additional 'from' expression.
    -- GHC **knows** that in this case 'a == String'.
    Just Refl -> x ++ "!!"
    Nothing ->
      case eqT @a @b of
        -- It's a proof for GHC which allows
        -- to apply 'from x' expression.
        -- GHC **knows** that in this case 'a == b' and
        -- there is 'instance From b String', which means
        -- that there is 'instance From a String'.
        Just Refl -> from x ++ "!!"
        Nothing ->
          error $
            "This will never happen"
              <> " because a == String"
              <> " and a == b"
              <> " cover all possibilities."

str :: String
str = appendBangBang ("hello" :: String)

txt :: String
txt = appendBangBang ("world" :: Text)
```

**Comment by tfausak** [comment]
That is neat! Unfortunately I wouldn't suggest that to users of Witch as a reasonable workaround for the lack of a `From a a` instance. And even if it was simplified to something like [`cast`](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-Typeable.html#v:cast), you'd still need both the `Typeable` constraint and the `From` constraint.

**Comment by 21it** [comment]
`Typeable` instances are not that big deal, because they are always there provided for free, there is no need to derive or implement them. The biggest problem I do see with this solution is that it does not look ergonomic. Pattern matching on `Refl` is ok, but usage of `Void` and `error` is ugly despite the fact that in reality such `error` placeholders will never be applied. GHC compiles code and it works fine, but another problem is that for some reason it gives a warning about "redundant" constraint `If (a == String) (Void ~ b) (a ~ b)` and I'm not sure why and how to fix that.

Another possible solution to solve unwanted implicit usage of `From a a` instance is just to create more restricted `from'` function, and use it instead of `from` where needed. It's reusing already existing `From a b` instance, so it's relatively cheap solution:

```haskell
from' ::                       
  forall source target.        
  ( From source target,        
    'False ~ (source == target)
  ) =>                         
  source ->                    
  target                       
from' = from                   
```

The only issue there is GHC warning about redundant constraint `'False ~ (source == target)`, but maybe it can be fixed somehow, or maybe `from'` function can be located in a separate file with `-Wno-redundant-constraints` GHC option, to keed all other source code clean from redundant constraints.

**Comment by tfausak** [comment]
I think the `From a a` instance should remain. If you want to avoid it, you could define and use a function like `from'` that @21it provided. Ideally the Witch library could provide a way to notice these unnecessary conversions and warn about them, but that doesn't seem to be easily possible.

**Comment by 21it** [comment]
I did tried to re-import combinators with additional inequality restrictions to exclude `From a a` and potential `TryFrom a a` instances, and it worked surprisingly well https://github.com/21it/src/blob/6dfee5d991536ddc7d2415058469e30eda6c1344/bitfinex-client/src/BitfinexClient/Import/Witch.hs

As a result - compiler found all redundant `from` expressions for me, and it's quite convenient to re-import and use restricted combinators with the same names (in most places it works just fine from the box with restrictions). Maybe it make sense to add restricted combinators in a separate module of WItch itself, to have 2 kind of possible module imports, similar to `Data.Map.Lazy` and `Data.Map.Strict`. One additional thing I noticed - maybe it make sense to add inequality restriction to `TryFrom` class, to prevent even possibility of `TryFrom a a` instance which violates Witch laws because `From a a` instance already exist. Module `Data.Type.Equality` is part of the base, so no additional dependency footprint there I guess.

```haskell
import Data.Type.Equality (type (==))

class                                                  
  ( 'False ~ (source == target)                        
  ) =>                                                 
  TryFrom source target                                
  where                                                
  tryFrom ::                                           
    source ->                                          
    Either                                             
      ( TryFromException.TryFromException source target
      )                                                
      target                                           
```
